concepts/box/about.md

```diff
@@ -1,3 +1,153 @@
+}
+
+// `match` arms have incompatible types
+// expected struct `WerewolfOfLondon`, found struct `Rooster`
+// this is found to be of type `WerewolfOfLondon`
+// `match` arms have incompatible types
+// you could change the return type to be a boxed trait object
+// if you change the return type to expect trait objects, box the returned expressions
+
+```
+
+The compiler, in all of its helpfulness, tells us exactly what to do, which is to return a boxed trait object. If you don't know what a trait object is,
+a simple way to describe it is a way of returning a specific object via [dynamic dispatch][dynamic dispatch] from a function that could return any object
+implementing a specified trait. Since we need dynamic dispatch, we can try just putting a `dyn` in front of the trait name and see if we get away with it.
+
+```rust
+fn annoy_me(time_of_day: TimeOfDay) -> dyn Annoyance {
+    match time_of_day {
+        TimeOfDay::Midnight => WerewolfOfLondon {},
+        TimeOfDay::Morning => Rooster {},
+    }
+}
+
+// the size for values of type `dyn Annoyance` cannot be known at compilation time
+// the trait `Sized` is not implemented for `dyn Annoyance`
+// all local variables must have a statically known size
+// unsized locals are gated as an unstable feature
+```
+
+One of the errors is that the size for `dyn Annoyance` cannot be known at compile time. This is because annoy_me does not know what specific trait object
+it will return in advance of when it is called. Structs implementing Annoyance could have different fields and be different sizes. And we are reminded that
+one of the uses for `Box` is for a type whose size can’t be known at compile time. So now we try using a `Box`ed return value, and we `Box`the values returned
+from the match expression, like so
+
+```rust
+fn annoy_me(time_of_day: TimeOfDay) -> Box<dyn Annoyance> {
+    match time_of_day {
+        TimeOfDay::Midnight => Box::new(WerewolfOfLondon {}),
+        TimeOfDay::Morning => Box::new(Rooster {}),
+    }
+}
+```
+
+The signature for `Box` is `Box<T>` where `T` stands for the specific type to be boxed which is placed within the angle brackets, as in `Box<dyn Annoyance>` above.
+
+We can consume the boxed trait object like so
+
+```rust
+fn be_annoying(annoying_one: &dyn Annoyance) {
+    println!("{:?}", annoying_one.annoy());
+}
+
+pub fn main() {
+    let my_annoyance = annoy_me(TimeOfDay::Midnight);
+    be_annoying(&*my_annoyance);
+    let my_annoyance = annoy_me(TimeOfDay::Morning);
+    be_annoying(&*my_annoyance);
+}
+```
+
+which prints
+
+```rust
+"ahhh-OOOOOOO!"
+"errr-errr-errr-errr-ERRRRRRRRR!"
+```
+
+Since my_annoyance is of type `Box<dyn Annoyance>` it must be dereferenced from the `Box` using the deref operator `*`. It is then referenced as a
+`dyn Annoyance` for the be_annoying function to borrow it.
+
+Another way to implement it would be for the be_annoying function to accept a `Box<dyn Annoyance>` and use [deref coercion][deref coercion] to avoid
```

Comment:
A third way:

```rust
fn be_annoying<A: Annoyance + ?Sized>(annoying_one: &A) {
    println!("{:?}", annoying_one.annoy());
}
```

[playground](https://play.rust-lang.org/?version=stable&mode=debug&edition=2018&gist=cc6f839bfb2bb6acf281baf656972dd9)

This uses static instead of dynamic dispatch; the compiler has monomorphized `fn be_annoying` in this case, and called the appropriate variant for each variant of `Annoyance`.